internal/probes/main.go

```diff
@@ -0,0 +1,224 @@
+			return false
+		}
+
+		timeUntilProbe = time.Until(probe.Status.LastCheckedAt.Time.Add(probe.Spec.Interval.Duration))
+		if timeUntilProbe <= 0 {
+			logger.V(1).Info("probe ready to execute", "timeUntilProbe", timeUntilProbe, "lastChecked", probe.Status.LastCheckedAt, "interval", probe.Spec.Interval, "now", time.Now())
+			break
+		}
+
+		logger.V(1).Info("sleeping until probe is ready", "timeUntilProbe", timeUntilProbe, "lastChecked", probe.Status.LastCheckedAt, "interval", probe.Spec.Interval, "now", time.Now())
+		time.Sleep(timeUntilProbe)
+	}
+
+	//if we slept for a while, refresh the probe from the api
+	if timeUntilProbe > time.Second*3 {
+		err := k8sClient.Get(ctx, client.ObjectKeyFromObject(originalProbe), probe)
+		if client.IgnoreNotFound(err) != nil {
+			logger.Error(err, "error reading probe")
+			return true
+		} else if err != nil {
+			logger.Info("could not find probe CR, killing related worker")
+			return false
+		}
+	}
+
+	result := performProbe(ctx, k8sClient, probe)
+
+	patchFrom := client.MergeFrom(probe.DeepCopy())
+	if !result.Healthy {
+		probe.Status.ConsecutiveFailures += 1
+	} else {
+		probe.Status.ConsecutiveFailures = 0
+	}
+	probe.Status.Healthy = &result.Healthy
+	probe.Status.LastCheckedAt = result.CheckedAt
+	probe.Status.Reason = result.Reason
+
+	err := k8sClient.Status().Patch(ctx, probe, patchFrom, &client.SubResourcePatchOptions{})
+	if err != nil {
+		logger.Error(err, "error patching probe status")
+	}
+	return true
+}
+
+func performProbe(ctx context.Context, k8sClient client.Client, probe *v1alpha1.DNSHealthCheckProbe) ProbeResult {
+	logger := log.FromContext(ctx)
+	logger.V(3).Info("performing health check", "request", probe)
+
+	//if address is a CNAME, check all IP Addresses that it resolves to
+	ips := probe.GetIPAddresses()
+
+	// Default port to 80
+	port := 80
+	if probe.Spec.Port != nil {
+		port = *probe.Spec.Port
+	}
+
+	// get any user-defined additional headers
+	headers, err := getAdditionalHeaders(ctx, k8sClient, probe)
+	if err != nil {
+		return ProbeResult{CheckedAt: metav1.Now(), Healthy: false, Reason: err.Error()}
+	}
+
+	for _, ip := range ips {
+		result := performRequest(ctx, string(probe.Spec.Protocol), probe.Spec.Hostname, probe.Spec.Path, ip, port, probe.Spec.AllowInsecureCertificate, headers)
+		// if any IP in a CNAME fails, it is a failed CNAME
+		if !result.Healthy {
+			return result
+		}
+	}
+
+	return ProbeResult{
+		CheckedAt: metav1.Now(),
+		Healthy:   true,
+	}
+}
+
+func performRequest(ctx context.Context, protocol, host, path, ip string, port int, allowInsecure bool, headers v1alpha1.AdditionalHeaders) ProbeResult {
+	probeClient := instrumented_http.NewClient(&http.Client{
+		Transport: TransportWithDNSResponse(map[string]string{host: ip}, allowInsecure),
+	}, &instrumented_http.Callbacks{
+		PathProcessor: func(path string) string {
+			parts := strings.Split(path, "/")
+			return parts[len(parts)-1]
+		},
+	})
```

Comment:
I removed the dependency on `github.com/linki/instrumented_http` in favour of just using the prometheus http middleware in [this PR]( https://github.com/Kuadrant/dns-operator/pull/244). Consider using [this](https://github.com/Kuadrant/dns-operator/blob/main/internal/metrics/http.go#L39) instrumented http client instead.

Reply:
If you do intend to use this, you will also need to register it with the controller runtime metrics somewhere:

```
metrics.Registry.MustRegister(instrumented_http.RequestDurationSeconds)
```

Reply:
I've updated it as you suggested, so I don't need to do that now, right?

Reply:
> I've updated it as you suggested, so I don't need to do that now, right?

No

Reply:
This will create a single set of metrics for all probes which might not be that useful. I'd go with it just now though, we can always modify the client to make the metrics it's producing more configurable since presumably you would want them keyed off the host/path/ip to be anyway useful.

Reply:
yes, I considered adding a host and endpoint to the name, but when I looked at your names they were all quite short, and I wasn't sure if they needed to stay short or not.